File: Programmars/Level_2/오픈채팅방.py
def solution(record):
    answer = []
    commands = []
    uids = []
    info = {}

    for rec in record:
        rec = rec.split()
        if len(rec) == 3:  # Enter, Change
            command, uid, nick = rec
            info[uid] = nick
        else:  # Leave
            command, uid = rec

        if command == 'Enter' or command == 'Leave':
            commands.append(command)
            uids.append(uid)
        else:
            info[uid] = nick

    for i in range(len(commands)):
        if commands[i] == 'Enter':
            answer.append(f'{info[uids[i]]}님이 들어왔습니다.')
        else:
            answer.append(f'{info[uids[i]]}님이 나갔습니다.')
    return answer


print(solution(["Enter uid1234 Muzi", "Enter uid4567 Prodo",
                "Leave uid1234", "Enter uid1234 Prodo", "Change uid4567 Ryan"]))  # ["Prodo님이 들어왔습니다.", "Ryan님이 들어왔습니다.", "Prodo님이 나갔습니다.", "Prodo님이 들어왔습니다."]

File: JavaScript/Programmars/카카오블라인드/오픈채팅방.js
function solution(record) {
  const userInfo = {};
  const actions = [];
  const messages = {
    Enter: '님이 들어왔습니다.',
    Leave: '님이 나갔습니다.',
  };

  for (let str of record) {
    const [action, uid, nick] = str.split(' ');
    if (action === 'Enter' || action === 'Leave') {
      actions.push([action, uid]);
    }
    if (action === 'Enter' || action === 'Change') {
      userInfo[uid] = nick;
    }
  }

  return actions.map(([action, uid]) => `${userInfo[uid]}${messages[action]}`);
}

console.log(
  solution([
    'Enter uid1234 Muzi',
    'Enter uid4567 Prodo',
    'Leave uid1234',
    'Enter uid1234 Prodo',
    'Change uid4567 Ryan',
  ])
); // ["Prodo님이 들어왔습니다.", "Ryan님이 들어왔습니다.", "Prodo님이 나갔습니다.", "Prodo님이 들어왔습니다."]
